swap/src/cli/watcher.rs

```diff
@@ -0,0 +1,104 @@
+use std::sync::Arc;
+use std::collections::HashMap;
+use std::time::Duration;
+
+use anyhow::Result;
+use uuid::Uuid;
+
+use crate::bitcoin::wallet::ScriptStatus;
+use crate::bitcoin::{ExpiredTimelocks, TxLock, Wallet};
+use crate::cli::api::tauri_bindings::TauriHandle;
+use crate::protocol::bob::BobState;
+use crate::protocol::{Database, State};
+
+use super::api::tauri_bindings::TauriEmitter;
+
+/// A long running task which watches for changes to timelocks and the number of confirmations.
+#[derive(Clone)]
+pub struct Watcher {
+    wallet: Arc<Wallet>,
+    database: Arc<dyn Database + Send + Sync>,
+    /// This saves for every running swap the expired timelocks as well as
+    /// the [`ScriptStatus`] of [`TxLock`].
+    current_swaps: HashMap<(Uuid, TxLock), (ExpiredTimelocks, ScriptStatus)>,
+    tauri: Option<TauriHandle>,
+}
+
+impl Watcher {
+    /// How often to check for changes (in seconds)
+    const CHECK_INTERVAL: u64 = 3;
+
+    /// Create a new Watcher
+    pub fn new(wallet: Arc<Wallet>, database: Arc<dyn Database + Send + Sync>, tauri: Option<TauriHandle>) -> Self {
+        Self {
+            wallet,
+            database,
+            current_swaps: HashMap::new(),
+            tauri,
+        }
+    }
+
+    /// Start running the watcher event loop. 
+    /// Should be done in a new task using [`tokio::spawn`].
+    pub async fn run(mut self) {
+        // Note: since this is de-facto a daemon, we have to gracefully handle errors
+        // (which in our case means logging the error message and trying again later)
+        loop {
+            // Fetch current transactions and timelocks
+            let current_swaps = match self.get_current_swaps().await {
+                Ok(val) => val,
+                Err(e) => {
+                    tracing::error!(error=%e, "Failed to fetch current transactions, retrying later");
+                    continue;
+                }
+            };
+
+            // Check for changes for every current swap
+            for (uuid, state) in current_swaps {
+                // Check if the timelock has expired
+                let new_timelock_status = match state.expired_timelocks(self.wallet.clone()).await {
+                    Ok(Some(val)) => val,
+                    Ok(None) => continue, // ignore finished swaps
+                    Err(e) => {
+                        tracing::error!(error=%e, "Failed to fetch expired timelocks, retrying later");
+                        continue;
+                    }
+                };
+                let new_confirmation_status = match self.wallet.status_of_script(state.tx_lock()) {
```

Comment:
I believe what you need to do here is subscribe instead?

Additionally, you need to subscribe the tx_cancel as well. Check the expired_timelock function for which transactions need to be watched.

Reply:
I don't see how we could subscribe here, given that that would require awaiting which we can't really do here. Since we check every 3 secs we should just check the status directly.

Reply:
I got it now. You are checking the confirmation of `tx_lock` to display them in the GUI while the user is waiting for Alice to lock their Monero.

I think it's better to only use the `watcher` for watching the timelocks (which we need to do contineously throughout all stages of a pending swap) while we watch for the confirmations of the Bitcoin lock transaction in the state machine, similiar to how it's done for the Monero lock transaction (because we only care about the confirmations of `tx_lock` while we are in the `BtcLocked` state. Afterwards, all we are about is the status of the timelock).